tools/pytorch/run.py

```diff
@@ -20,9 +21,19 @@ def tensor(x):
 
 
 def output(ret):
-    if ret.size() == 1:
-        return ret.item()
-    return ret.tolist()
+    try:
+        if type(ret) == BASparseMat:
+            return str(f"BASparseMat: {ret.nrows} rows by {ret.ncols} columns")
+        if type(ret) == tuple:
+            reproj, w = ret
+            len_r = len(reproj.tolist())
+            len_w = len(w.tolist())
+            return f"[{len_r/w}{reproj.tolist()[:2]},{len_w}{w.tolist()[0]}]"
```

Comment:
Instead of converting these to strings, can you make them into structured data? Similar to how the GMM `input` is a JSON object with multiple fields containing different parts of the data.

It'd probably make sense to do it similar to how we're doing it in DiffSharp, where each function (like `calculate_objectiveBA` and `calculate_jacobianBA`) can have its own functions attached, to convert the input from JSON and the output to JSON.

Reply:
Sounds good! Do you mean something like `{"BASparseMat": {"rows": mat.nrows, "columns": mat.ncols }}` rather than the string? And `{"reproj_error": r_err.tolist(), "w_err": w_err.tolist()}`? For reproj and w error, they are huge lists where the first element (or two) are just repeated a tone of times. Do you still want me to do something like that to make it viewable or leave the whole list?

Reply:
Yep that sounds perfect! Since you've mentioned that the error vectors are just the same numbers duplicated many times, you could just output the data point and also how many times it's duplicated. And for the Jacobian matrix, for now it's fine to just output the dimensions; maybe later we'll decide to change it so we serialize the sparse matrix data, but no need to do that right now.